### ISSUE TITLE
Running as system user with no /home/jackett directory

### ISSUE THREAD

**Issue by DominicMe** [Jackett/Jackett#1469]
I am trying to jun jacket on Centos 7 as system user but I am getting errors relating to the home directory:
`
 Could not create settings directory. Access to the path "/home/jackett" is denied.`

This happens when attempting to add trackers with user/pass otherwise .json file is created just fine in the data dir as specified.

Here how I execute jackett:
`sudo -u jackett /usr/bin/mono /opt/jackett/JackettConsole.exe -d /opt/jackett/data`

Why is jackett attempting to use home directory when data is set?? The odd thing is that this exact setup worked fine an Arch linux with no issues. What am I missing?

Jackett version:  4.2 and same with latest

-------------------------------

**Comment by flightlevel** [comment]
Can you try a newer Mono version first 4.8 or 5? I know 4.2 had some bugs, not sure it would do this though ...

**Comment by DominicMe** [comment]
Yes, that's the first thing I did. I am now using version 5.0.1.1 which is the latest stable release. I have as far as I can tell duplicated the setup from my Arch Install but still it fails on startup due to no permissions to non-existing /home/jackett directory. Is it hard coded somewhere by any chance? How is the startup directory chosen?

**Comment by flightlevel** [comment]
https://github.com/Jackett/Jackett/blob/f2def1f615a2df8cb929860992c22aec95906c4f/src/Jackett/Services/ConfigurationService.cs#L245

What path does it say in the logs?
https://github.com/Jackett/Jackett/blob/f2def1f615a2df8cb929860992c22aec95906c4f/src/Jackett.Console/Program.cs#L107

'
This happens when attempting to add trackers with user/pass otherwise .json file is created just fine in the data dir as specified.
'
Can you explain this bit more clearly, the json file created in the data directory just fine ???
So you are getting that above error, but everything is working okay????
Where is the log file being stored?

https://github.com/Jackett/Jackett/blob/f2def1f615a2df8cb929860992c22aec95906c4f/src/Jackett/Services/ConfigurationService.cs 
You can use the above to assist in debugging

**Comment by DominicMe** [comment]
It was starting before even with errors relating to the /home/jackett dir but that was before mono update and possibly some other change.

Here is the command I run:

`sudo -u jackett /usr/bin/mono /opt/jackett/JackettConsole.exe -d /opt/jackett/data`

I was about to post the error log but it actually did start again with no errors. Here is the log showing what happens when I attempt to add an indexer with user/pass. Everything after the blank line is what happens when I hit "Okay".

```
06-12 11:34:42 Info Adding aggregate indexer 
06-12 11:34:42 Info Starting web server at http://*:9117/ 
06-12 11:35:06 Info redirecting to /Admin/Dashboard 

06-12 11:36:19 Error Exception in Configure System.Security.Cryptography.CryptographicException: Data protection failed. ---> System.Security.Cryptography.CryptographicException: Could not create user key store '/home/jackett/.config/.mono/keypairs'. ---> System.UnauthorizedAccessException: Access to the path "/home/jackett" is denied.
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x0005e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.get_UserPath () [0x0005e] in <3753d1715b8842d8bb13a30db0388b60>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Cryptography.KeyPairPersistence.get_UserPath () [0x00086] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.get_Filename () [0x00063] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.Load () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Security.Cryptography.RSACryptoServiceProvider.Common (System.Security.Cryptography.CspParameters p) [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Security.Cryptography.RSACryptoServiceProvider..ctor (System.Int32 dwKeySize, System.Security.Cryptography.CspParameters parameters) [0x0001d] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.ManagedProtection.GetKey (System.Security.Cryptography.DataProtectionScope scope) [0x00034] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at Mono.Security.Cryptography.ManagedProtection.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x00133] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at System.Security.Cryptography.ProtectedData.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x00023] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
   --- End of inner exception stack trace ---
  at System.Security.Cryptography.ProtectedData.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x0003f] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at Jackett.Services.ProtectionService.ProtectDefaultMethod (System.String plainText) [0x0006a] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Services.ProtectionService.Protect (System.String plainText) [0x00011] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Models.IndexerConfig.ConfigurationData.ToJson (Jackett.Services.IProtectionService ps, System.Boolean forDisplay) [0x0013d] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Indexers.BaseIndexer.SaveConfig () [0x00000] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Indexers.BaseIndexer+<ConfigureIfOK>c__asyncC.MoveNext () [0x00049] in <548275e367574f8798abfe134b3e9606>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Jackett.Indexers.Demonoid+<ApplyConfiguration>c__async0.MoveNext () [0x001eb] in <548275e367574f8798abfe134b3e9606>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Jackett.Controllers.AdminController+<Configure>c__async4.MoveNext () [0x00167] in <548275e367574f8798abfe134b3e9606>:0 


```

**Comment by kaso17** [comment]
This seems to be a mono specific issue.
I'm not sure if it can be fixed.
Seems like this code is responsible for getting the UserPath for the keystore. I'm not sure if we can override it.
https://github.com/mono/mono/blob/5df45ecd4e8083c450013ad3432943cc4c2ed25e/mcs/class/Mono.Security/Mono.Security.Cryptography/KeyPairPersistence.cs#L206-L241

I would suggest that you try to change the HOME directory of your jackett user to e.g. "/opt/jackett/". Then you probably won't need the -d argument at all.

**Comment by DominicMe** [comment]
I am using 5.0.1.1 on centos and 5.0.0.100 on Arch. _userPath is not hard-coded in mono because 1. it works on Arch for whatever reason and 2. as you say /opt/jackett should work. This means it's not an issue with mono but with Jackett. If /opt/jackett works (and it might be the one difference between my arch install and centos install) but -d /some/dir does not, then the issue is that somewhere (but not everywhere) Jackett fails to pass -d option properly.

**Comment by DominicMe** [comment]
Did more testing an a clean VM. Jackett fails to start with no -d option which makes sense since /home/dir must be the default dir when it's not specified.

When started with -d option it starts with 0 errors but fails with Data Protection error when trying to add an indexer with a password.

```
06-15 15:06:37 Error Exception in Configure System.Security.Cryptography.CryptographicException: Data protection failed. ---> System.Security.Cryptography.CryptographicException: Could not create user key store '/home/jackett/.config/.mono/keypairs'. ---> System.UnauthorizedAccessException: Access to the path "/home/jackett" is denied.
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x0005e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.DirectoryInfo.Create () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo:Create ()
  at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x00036] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.get_UserPath () [0x0005e] in <3753d1715b8842d8bb13a30db0388b60>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Cryptography.KeyPairPersistence.get_UserPath () [0x00086] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.get_Filename () [0x00063] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.KeyPairPersistence.Load () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Security.Cryptography.RSACryptoServiceProvider.Common (System.Security.Cryptography.CspParameters p) [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Security.Cryptography.RSACryptoServiceProvider..ctor (System.Int32 dwKeySize, System.Security.Cryptography.CspParameters parameters) [0x0001d] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Mono.Security.Cryptography.ManagedProtection.GetKey (System.Security.Cryptography.DataProtectionScope scope) [0x00034] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at Mono.Security.Cryptography.ManagedProtection.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x00133] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at System.Security.Cryptography.ProtectedData.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x00023] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
   --- End of inner exception stack trace ---
  at System.Security.Cryptography.ProtectedData.Protect (System.Byte[] userData, System.Byte[] optionalEntropy, System.Security.Cryptography.DataProtectionScope scope) [0x0003f] in <d0647575f06b4d40805ba6e50fc9fbd2>:0 
  at Jackett.Services.ProtectionService.ProtectDefaultMethod (System.String plainText) [0x0006a] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Services.ProtectionService.Protect (System.String plainText) [0x00011] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Models.IndexerConfig.ConfigurationData.ToJson (Jackett.Services.IProtectionService ps, System.Boolean forDisplay) [0x0013d] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Indexers.BaseIndexer.SaveConfig () [0x00000] in <548275e367574f8798abfe134b3e9606>:0 
  at Jackett.Indexers.BaseIndexer+<ConfigureIfOK>c__asyncC.MoveNext () [0x00049] in <548275e367574f8798abfe134b3e9606>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Jackett.Indexers.Demonoid+<ApplyConfiguration>c__async0.MoveNext () [0x001eb] in <548275e367574f8798abfe134b3e9606>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <3753d1715b8842d8bb13a30db0388b60>:0 
  at Jackett.Controllers.AdminController+<Configure>c__async4.MoveNext () [0x00167] in <548275e367574f8798abfe134b3e9606>:0 

```

This is clearly a bug. Seems like the data directory path is not being passed to mono somewhere. This means that the -d option is broken and unusable. I hope it get's fixed as in my case I can't really use it at all due this bug.

**Comment by kaso17** [comment]
Please try setting the XDG_CONFIG_HOME environment variable to the data dir.
Just changing the users homedir still seems the simpler workaround (`usermod -m -d  /opt/jackett username`)

**Comment by DominicMe** [comment]
It's a system user so it has no home directory. In any case, /opt directory is the wrong place to install a system service. The correct solution is of course a fix to this bug but unfortunately I am not familiar with the code to attempt it.

**Comment by kaso17** [comment]
@DominicMe did you try setting XDG_CONFIG_HOME?

**Comment by DominicMe** [comment]
Did a clean Jackett install on a VM again this time running export XDG_CONFIG_HOME=/var/lib/jackett before Jackett. Results are the same, it still attempts to use /home/jackett. /opt is the wrong place to install Jackett, and Both Sonarr and Radarr use mono too with no such issues. It's a serious bug and should be simple to to fix not to mention to having to deal with continuous issues people have because of this.

**Comment by kaso17** [comment]
I looked into this.
Setting XDG_CONFIG_HOME correctly is the only way to make it work.
I tested it with mono 5 and it works as expected:
```
$ XDG_CONFIG_HOME=/tmp/test mono /home/jackett/Jackett/JackettConsole.exe
08-29 14:00:37 Info App config/log directory: /tmp/test/Jackett
```
Setting XDG_CONFIG_HOME from the inside doesn't affect the DataProtection class path selection. It must be set before running mono.

Relevant code:
https://github.com/mono/mono/blob/master/mcs/class/Mono.Security/Mono.Security.Cryptography/KeyPairPersistence.cs#L210-L212

I've update the DataDir description via 300354ef13688e338bea68b79cc278ede1d089be
